ipa-core/src/protocol/basics/share_validation.rs

```diff
@@ -0,0 +1,166 @@
+use futures_util::future::try_join;
+
+use crate::{
+    error::Error,
+    helpers::{
+        hashing::{compute_hash, Hash},
+        Direction,
+    },
+    protocol::{context::Context, RecordId},
+    secret_sharing::SharedValue,
+};
+
+/// This function checks that a vector of shares are consistent across helpers
+/// i.e. `H1` holds `(x0,x1)`, `H2` holds `(x1,x2)`, `H3` holds `(x2,x0)`
+/// the function verifies that `H1.x0 == H3.x0`, `H1.x1 == H2.x1`, `H2.x2 == H3.x2`
+///
+/// We use a hash based approach that is secure in the random oracle model
+/// further, only one of left and right helper check that it is zero
+/// this is might not be sufficient in some applications to prevent malicious behavior
+///
+/// The left helper simply hashes the vector and sends it to the right,
+/// the right helper negates his vector, hashes it and compares it to the received hash
+///
+/// # Errors
+/// propagates errors from send and receive
+pub async fn validate_replicated_shares<'a, 'b, C, I, J, S>(
+    ctx: C,
+    input_left: I,
+    input_right: J,
+) -> Result<(), Error>
+where
+    C: Context,
+    I: IntoIterator<Item = &'a S>,
+    J: IntoIterator<Item = &'b S>,
+    S: SharedValue,
+{
+    // compute hash of `left`
+    let hash_left = compute_hash(input_left);
+
+    // set up context
+    let ctx_new = &(ctx.set_total_records(1usize));
+    // set up channels
+    let send_channel = ctx_new.send_channel::<Hash>(ctx.role().peer(Direction::Right));
+    let receive_channel = ctx_new.recv_channel::<Hash>(ctx.role().peer(Direction::Left));
+
+    let ((), hash_right) = try_join(
```

Comment:
I would recommend naming this something different. This name is super confusing. Can we call it `hash_received` instead?